validator/coordinator/src/main/java/tech/pegasys/teku/validator/coordinator/performance/RecentChainDataPerformanceTracker.java

```diff
@@ -0,0 +1,261 @@
+    Set<BeaconBlock> blockInEpoch = getBlocksInEpochs(startEpochInclusive, endEpochExclusive);
+    List<SignedBeaconBlock> sentBlocks =
+        sentBlocksByEpoch.subMap(startEpochInclusive, true, endEpochExclusive, false).values()
+            .stream()
+            .flatMap(Collection::stream)
+            .collect(Collectors.toList());
+    long numberOfIncludedBlocks =
+        sentBlocks.stream()
+            .map(SignedBeaconBlock::getMessage)
+            .filter(blockInEpoch::contains)
+            .count();
+
+    return new BlockPerformance((int) numberOfIncludedBlocks, sentBlocks.size());
+  }
+
+  private AttestationPerformance getAttestationPerformanceForEpoch(
+      UInt64 currentEpoch, UInt64 analyzedEpoch) {
+    checkArgument(
+        analyzedEpoch.isLessThanOrEqualTo(currentEpoch.minus(UInt64.valueOf(2))),
+        "Epoch to analyze attestation performance must be at least 2 epochs less than the current epoch");
+    // Attestations can be included in either the epoch they were produced in or in
+    // the following epoch. Thus, the most recent epoch for which we can evaluate attestation
+    // performance is current epoch - 2.
+    UInt64 analysisRangeEndEpoch = analyzedEpoch.plus(UInt64.valueOf(2));
+
+    // Get included attestations for the given epochs in a map from slot to attestations
+    // included in block.
+    Map<UInt64, List<Attestation>> attestations =
+        getAttestationsIncludedInEpochs(analyzedEpoch, analysisRangeEndEpoch);
+
+    // Get sent attestations in range
+    Set<Attestation> sentAttestations =
+        sentAttestationsByEpoch.getOrDefault(analyzedEpoch, new HashSet<>());
+    UInt64 analyzedEpochStartSlot = compute_start_slot_at_epoch(analyzedEpoch);
+    UInt64 rangeEndSlot = compute_start_slot_at_epoch(analysisRangeEndEpoch);
+    BeaconState state = recentChainData.getBestState().orElseThrow();
+
+    int correctTargetCount = 0;
+    int correctHeadBlockCount = 0;
+    List<Integer> inclusionDistances = new ArrayList<>();
+
+    for (Attestation sentAttestation : sentAttestations) {
+      // Check if the sent attestation is included in any block in the appropriate range.
+      // Appropriate range being: [ attestation_production_epoch, attestation_production_epoch + 1 ]
+      UInt64 attestationSlot = sentAttestation.getData().getSlot();
+      for (UInt64 currSlot = analyzedEpochStartSlot;
+          currSlot.isLessThan(rangeEndSlot);
+          currSlot = currSlot.increment()) {
+        if (attestations.containsKey(currSlot)) {
+          if (checkIfAttestationIsIncludedInList(sentAttestation, attestations.get(currSlot))) {
+            inclusionDistances.add(currSlot.minus(attestationSlot).intValue());
+          }
+        }
+      }
+
+      // Check if the attestation had correct target
+      Bytes32 attestationTargetRoot = sentAttestation.getData().getTarget().getRoot();
+      if (attestationTargetRoot.equals(get_block_root(state, analyzedEpoch))) {
```

Comment:
We already have a utility `recentChainData.getBlockRootBySlot` - we could add a similar one for checking the root at an epoch boundary:
```suggestion
      final Optional<Bytes32> rootAtEpoch = recentChainData.getBlockRootByEpoch(state, analyzedEpoch);
      if (rootAtEpoch.map(r -> r.equals(attestationTargetRoot).orElse(false)) {
```

Although, maybe we could run into an issue with data getting pruned out from under us?

Reply:
This is not an option anymore since we switched to CombinedChainDataClient.